net: eth: Convert to use callbacks to query stats

jyong2 opened the pull request:
The advantage to this approach allows drivers for
devices that already keep statistics data on hardware
registers to use those instead, rather than try to
replicate it the same counters again within the driver
itself.

The eth_native_posix.c driver though do not benefit
from this, is modified to use the new callback system.

@jukkar, @anashif

pfalcon reviewed (dismissed):
This patch is described as "net: eth: Convert to use callbacks to query stats", but actually does much more than that, for example, effectively eliminates entire API introduced in c1e7fd76ef0fb351f720bf7793eea387342a6da6, file subsys/net/l2/ethernet/eth_stats.h, while leaving that effectively empty file around.

At the very least, the commit message should mention that this is being done, and explain why. Even better is to have 2 commits, first one doing nothing but introducing the callback, the second one removing the existing API with the reasons given. (And of course, with no stale files left around.)

therealprof commented on subsys/net/l2/ethernet/ethernet_stats.c:
```diff
@@ -29,7 +29,7 @@ static int eth_stats_get(u32_t mgmt_request, struct net_if *iface,
 		eth = net_if_get_device(iface)->driver_api;
 		len_chk = sizeof(struct net_stats_eth);
-		src = eth->stats;
+		src = eth->get_stats(iface);
```
Missing a check whether the driver implements `get_stats()` here, otherwise it'll crash.
```
+         if (eth->get_stats == NULL) {
+                 return -ENOENT;
+         }
```

jyong2 replied:
Thanks for catching that.

therealprof reviewed (changes requested):
LGTM and works fine (except for the missing counters) on a Nucleo-F207ZG with ethernet interface after adding the NULL pointer check; without that it'll crash when requesting interface specific stats.

jyong2 commented:
New version pushed, it leaves the existing inline functions in subsys/net/l2/ethernet/eth_stats.h, but adapts them to call get_stats.

therealprof commented on subsys/net/l2/ethernet/eth_stats.h:
```diff
@@ -17,9 +17,13 @@ static inline void eth_stats_update_bytes_rx(struct net_if *iface,
 
-	stats = ((const struct ethernet_api *)
-		 net_if_get_device(iface)->driver_api)->stats;
+	if (!api->get_stats)
```
Missing {}

jyong2 replied:
Done, updated in dcbadc52a66f1e31e9dbe081fa4aeb2e6c36a727.

jukkar commented on drivers/ethernet/eth_native_posix.c:
```diff
@@ -441,14 +441,23 @@ static struct device *eth_get_ptp_clock(struct device *dev)
 
 #if defined(CONFIG_NET_STATISTICS_ETHERNET)
-	.stats = &eth_context_data.stats,
+	.get_stats = &get_stats,
```
Function name is already its address and you should not use & here. Just do `.get_stats = get_stats,`

jyong2 replied:
Done.

jukkar commented on subsys/net/l2/ethernet/eth_stats.h:
```diff
@@ -70,9 +89,13 @@ static inline void eth_stats_update_pkts_tx(struct net_if *iface)
 
-	stats = ((const struct ethernet_api *)
-		 net_if_get_device(iface)->driver_api)->stats;
+	if (!api->get_stats)
```
Please make sure that all the one line "if" statements have {}

jyong2 replied:
Sorry, missed this one.

jyong2 commented:
Updated names in commit message.

jyong2 commented:
The latest error reads:
e33bd0b55f57ef97b2be997d1db92a334386185f: author email does not follow the syntax: First Last <email>.

Where is that commit from?

jyong2 committed:
net: eth: Convert to use callbacks to query stats

The advantage to this approach allows drivers for
devices that already keep statistics data on hardware
registers to use those instead, rather than try to
replicate it the same counters again within the driver
itself.

The eth_native_posix.c driver though do not benefit
from this, is modified to use the new callback system.

Suggested-by: Jukka Rissanen <jukka.rissanen@intel.com>
Signed-off-by: Jonathan Yong <jonathan.yong@intel.com>
